File: templates/problems/06_linked_list/03_linked_list_cycle/python/solution.py
from helpers import ListNode, Optional


def solve_has_cycle(head: Optional[ListNode]) -> bool:
    """Detect a cycle using Floyd's tortoise and hare algorithm.

    Time: O(n), Space: O(1).
    """
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            return True
    return False

File: templates/problems/06_linked_list/03_linked_list_cycle/typescript/solution.ts
import { ListNode } from "./helpers";

export function solveHasCycle(head: ListNode | null): boolean {
    let slow = head;
    let fast = head;
    while (fast !== null && fast.next !== null) {
        slow = slow!.next;
        fast = fast.next.next;
        if (slow === fast) {
            return true;
        }
    }
    return false;
}
